crates/runtime/src/databackend/viewtable.rs

```diff
@@ -0,0 +1,98 @@
+use std::{pin::Pin, sync::Arc, thread, time::Duration};
+
+use datafusion::{
+    datasource::ViewTable,
+    error::DataFusionError,
+    execution::context::SessionContext,
+    sql::{parser::DFParser, sqlparser::dialect::AnsiDialect},
+};
+use futures_core::Future;
+use snafu::prelude::*;
+
+use super::{
+    DataBackend, DataBackendType, Result, UnableToAddDataSnafu, UnableToCreateTableDataFusionSnafu,
+    UnableToCreateTableSnafu, UnableToParseSqlSnafu, UnsupportedOperationSnafu,
+};
+
+pub struct ViewTableBackend {
+    _ctx: Arc<SessionContext>,
+    _name: String,
+    _sql: String,
+}
+
+impl ViewTableBackend {
+    pub async fn new(ctx: Arc<SessionContext>, name: &str, sql: &str) -> Result<Self> {
+        let statements = DFParser::parse_sql_with_dialect(sql, &AnsiDialect {})
+            .context(UnableToParseSqlSnafu)?;
+        if statements.len() != 1 {
+            return UnableToCreateTableSnafu {
+                reason: format!(
+                    "Expected 1 statement to create view, received {}",
+                    statements.len()
+                )
+                .to_string(),
+            }
+            .fail();
+        }
+
+        let view: ViewTable;
+        let mut backoff = Duration::from_secs(1);
+        let mut attempts = 1;
+        // Tables are currently lazily created (i.e. not created until first data is received) so that we know the table schema.
+        // This means that we can't create a view on top of a table until the first data is received for all dependent tables and therefore
+        // the tables are created. To handle this, if view creation fails with a plan error, we retry until the table is created.
+        loop {
+            let plan_result = ctx.state().statement_to_plan(statements[0].clone()).await;
+            match plan_result {
+                Ok(plan) => {
+                    view = ViewTable::try_new(plan, Some(sql.to_string()))
+                        .context(UnableToAddDataSnafu)?;
+                    break;
+                }
+                Err(e) => match e {
+                    DataFusionError::Plan(_) => {
```

Comment:
How can we distinguish between a view that is incorrect vs one that is referencing a dataset that is registered but doesn't have data yet?

Reply:
We can't, at least not without string matching and that is much more brittle. That's why for the time being I've implemented a backoff policy so that we eventually give up and assume it's a legitimate error.

Reply:
See my handoff for my thoughts on a longer term solution.

Reply:
Datafusion has a SQL parser we can use - that isn't string matching

Reply:
Yes, I understand that. Your comment here was on a line that was matching the errors. We can't get any more granular from the error itself without string matching.

See my comment below on SQL parsing + message publishing and subscribing: https://github.com/spiceai/spiceai/pull/653/files#r1479071322

Reply:
This has become a duplicate of the below comment, closing in favor of that one.